src/database.rs

```diff
@@ -34,64 +39,101 @@ pub struct Rows {
   pub rows_affected: Option<u64>,
 }
 pub type Headers = Vec<Header>;
-pub type DbPool = PgPool;
-pub type DbError = sqlx::Either<Error, ParserError>;
+pub type DbPool<DB> = Pool<DB>;
+pub type DbError = Either<Error, ParserError>;
+
+pub trait HasRowsAffected {
+  fn rows_affected(&self) -> u64;
+}
+
+pub trait DatabaseQueries {
+  fn preview_tables_query() -> String;
+  fn preview_rows_query(schema: &str, table: &str) -> String;
+  fn preview_columns_query(schema: &str, table: &str) -> String;
+  fn preview_constraints_query(schema: &str, table: &str) -> String;
+  fn preview_indexes_query(schema: &str, table: &str) -> String;
+  fn preview_policies_query(schema: &str, table: &str) -> String;
+}
+
+pub trait ValueParser: Database {
+  fn parse_value(row: &Self::Row, col: &Self::Column) -> Option<Value>;
+}
 
-pub async fn init_pool(opts: PgConnectOptions) -> Result<PgPool, Error> {
-  PgPoolOptions::new().max_connections(5).connect_with(opts).await
+pub trait BuildConnectionOptions: Database {
+  fn build_connection_opts(args: Cli) -> color_eyre::eyre::Result<<Self::Connection as Connection>::Options>;
+}
+
+pub async fn init_pool<DB: Database>(opts: <DB::Connection as Connection>::Options) -> Result<Pool<DB>, Error> {
+  PoolOptions::new().max_connections(3).connect_with(opts).await
 }
 
 // since it's possible for raw_sql to execute multiple queries in a single string,
 // we only execute the first one and then drop the rest.
-pub async fn query(query: String, pool: &PgPool) -> Result<Rows, DbError> {
-  let first_query = get_first_query(query);
+pub async fn query<DB>(query: String, dialect: &(dyn Dialect + Sync), pool: &Pool<DB>) -> Result<Rows, DbError>
+where
+  DB: Database + ValueParser,
+  DB::QueryResult: HasRowsAffected,
+  for<'c> &'c mut DB::Connection: Executor<'c, Database = DB>,
+{
+  let first_query = get_first_query(query, dialect);
   match first_query {
     Ok((first_query, _)) => {
       let stream = sqlx::raw_sql(&first_query).fetch_many(pool);
-      query_stream(stream).await
+      query_stream::<DB>(stream).await
     },
     Err(e) => Err(e),
   }
 }
 
-pub async fn query_stream(
-  mut stream: BoxStream<'_, Result<Either<PgQueryResult, PgRow>, Error>>,
-) -> Result<Rows, DbError> {
-  let mut query_finished = false;
+#[allow(clippy::type_complexity)]
+pub async fn query_stream<'a, DB>(
+  mut stream: BoxStream<'_, Result<Either<DB::QueryResult, DB::Row>, Error>>,
+) -> Result<Rows, DbError>
+where
+  DB: Database + ValueParser,
+  DB::QueryResult: HasRowsAffected,
+{
   let mut query_rows = vec![];
   let mut query_rows_affected: Option<u64> = None;
   let mut headers: Headers = vec![];
-  while !query_finished {
```

Comment:
I have to change this part as it hangs the MySQL connection. if I use 3 maximum connections, the third query will hang